tests/test_semver.py

```diff
@@ -0,0 +1,92 @@
+import unittest
+from collections import abc
+
+import packaging.version
+
+from imbi import semver
+
+
+class VersionCheckingTests(unittest.TestCase):
+    def verify_range(self, spec: str, in_range: abc.Iterable[str],
+                     out_of_range: abc.Iterable[str]) -> None:
+        ver_range = semver.parse_semver_range(spec)
+        for value in in_range:
+            self.assertIn(value, ver_range, value)
+        for value in out_of_range:
+            self.assertNotIn(value, ver_range, value)
+
+    def test_tilde_ranges_from_documentation(self) -> None:
+        # https://www.npmjs.com/package/semver#user-content-tilde-ranges-123-12-1
+
+        # ~1.2.3 := >=1.2.3 <1.(2+1).0 := >=1.2.3 <1.3.0-0
+        self.verify_range('~1.2.3', ('1.2.3', '1.2.4', '1.2.99'),
+                          ('1.1', '1.2.2', '1.2.2.99', '1.2', '1.3'))
+
+        # ~1.2 := >=1.2.0 <1.(2+1).0 := >=1.2.0 <1.3.0-0
+        self.verify_range('~1.2', ('1.2', '1.2.2', '1.2.3', '1.2.99'),
+                          ('1.1', '1.3.0', '1.3'))
+
+        # ~1 := >=1.0.0 <(1+1).0.0 := >=1.0.0 <2.0.0-0
+        self.verify_range('~1', ('1.2', '1.2.2', '1.2.3', '1.2.99', '1.99'),
+                          ('2.0', '0.1.2', '3'))
+
+        # ~0.2.3 := >=0.2.3 <0.(2+1).0 := >=0.2.3 <0.3.0-0
+        self.verify_range('~0.2.3', ('0.2.3', '0.2.99'),
+                          ('0.2.2', '0.3', '1.0'))
+
+        # ~0.2.3 := >=0.2.3 <0.(2+1).0 := >=0.2.3 <0.3.0-0
+        self.verify_range('~0.2.3', ('0.2.3', '0.2.99'),
+                          ('0.2.2', '0.3', '1.0'))
+
+        # ~0.2 := >=0.2.0 <0.(2+1).0 := >=0.2.0 <0.3.0-0
+        self.verify_range('~0.2.3', ('0.2.3', '0.2.99'),
+                          ('0.2.2', '0.3', '1.0'))
```

Comment:
These three tests are all the same. It looks like the second is a true dupe, and the third was supposed to test `~0.2` based on the comment.

Reply:
oops ... this is what happens when you reorganize your code half way through writing a test case.